packages/lib/passport-interface/src/PassportPopup/core.ts

```diff
@@ -0,0 +1,116 @@
+ */
+export async function zupassPopupSetup(): Promise<string | undefined> {
+  if (!window.opener) {
+    return "zupassPopupSetup() can only be called from within a popup window";
+  }
+
+  let params;
+
+  // Hash routing is commonly used in web applications to enable client-side
+  // routing without requiring server-side configuration, typically single-page applications.
+  // Without hash routing, the server should always serve the same index.html file for any route.
+  // Some providers, like Github Pages, don't provide this feature.
+  // To read the parameters of a URL with hash routing, the hash must first be removed.
+  if (window.location.href.includes(window.location.origin + "/#/")) {
+    const url = new URL(window.location.href.replace("#", ""));
+
+    params = url.searchParams;
+  } else {
+    params = new URLSearchParams(window.location.search);
+  }
+
+  const paramsProofUrl = params.get("proofUrl");
+  const paramsProof = params.get("proof");
+  const paramsEncodingPendingPCD = params.get("encodedPendingPCD");
+  const finished = params.get("finished");
+
+  // First, this page is window.open()-ed. Redirect to Zupass.
+  if (paramsProofUrl) {
+    window.location.href = paramsProofUrl;
+  } else if (finished) {
+    // Later, Zupass redirects back with a result. Send it to our parent.
+    if (paramsProof) {
+      window.opener.postMessage({ encodedPCD: paramsProof }, "*");
+    }
+
+    window.close();
+    await sleep(1000 * 3);
+    return "Finished. Please close this window.";
+  } else if (paramsEncodingPendingPCD) {
+    // Later, Zupass redirects back with a encodedPendingPCD. Send it to our parent.
+    window.opener.postMessage(
+      { encodedPendingPCD: paramsEncodingPendingPCD },
+      "*"
+    );
+    window.close();
+    await sleep(1000 * 3);
+    return "Finished. Please close this window.";
+  }
+}
+
+/**
+ * Open up a Zupass popup window using proofUrl from specific PCD integrations
+ * and popupUrl, which is the route where the useZupassPopupSetup hook is being
+ * served from.
+ */
+export function openZupassPopup(popupUrl: string, proofUrl: string): void {
+  const url = `${popupUrl}?proofUrl=${encodeURIComponent(proofUrl)}`;
+  window.open(url, "_blank", "width=450,height=600,top=100,popup");
+}
+
+/**
+ * The popup message can contain either an encoded PCD or encoded pending PCD.
+ */
+type PopupResultMessage =
+  | {
+      type: "pcd";
+      pcdStr: string;
+    }
+  | {
+      type: "pendingPcd";
+      pendingPcdStr: string;
+    };
+
+/**
+ * Asynchronous function that listens for PCDs and PendingPCDs from a Zupass
+ * popup window using message passing and event listeners.
+ */
+export function receiveZupassPopupMessage(): Promise<PopupResultMessage> {
+  return new Promise((resolve) => {
+    const receiveMessage = (ev: MessageEvent): void => {
+      if (ev.data.encodedPCD) {
+        console.log("Received PCD", ev.data.encodedPCD);
```

Comment:
These log messages were in the original. Not sure if they're still needed, so I'm leaving them in by default.

Reply:
I'd prefer to delete them. Zooming out, i'd like to re-consider how we do logging from a bird's eye view at some point. 

https://linear.app/0xparc-pcd/issue/0XP-589/better-logging-across-entire-app-suite